parla/cuda.py

```diff
@@ -163,27 +162,68 @@ def get_array_module(self, a):
 
 # Integration with parla.environments
 
+class _GPUStacksLocal(threading.local):
+    _stream_stack: List[cupy.cuda.Stream]
+    _device_stack: List[cupy.cuda.Device]
+
+    def __init__(self):
+        super(_GPUStacksLocal, self).__init__()
+        self._stream_stack = []
+        self._device_stack = []
+
+    def push_stream(self, stream):
+        self._stream_stack.append(stream)
+
+    def pop_stream(self) -> cupy.cuda.Stream:
+        return self._stream_stack.pop()
+
+    def push_device(self, dev):
+        self._device_stack.append(dev)
+
+    def pop_device(self) -> cupy.cuda.Device:
+        return self._device_stack.pop()
+
+    @property
+    def stream(self):
+        if self._stream_stack:
+            return self._stream_stack[-1]
+        else:
+            return None
+
+
 class GPUComponentInstance(EnvironmentComponentInstance):
-    stream: cupy.cuda.Stream
+    _stack: _GPUStacksLocal
     gpus: List[_GPUDevice]
 
     def __init__(self, descriptor: "GPUComponent", env: TaskEnvironment):
         super().__init__(descriptor)
         self.gpus = [d for d in env.placement if isinstance(d, _GPUDevice)]
         assert len(self.gpus) == 1
         self.gpu = self.gpus[0]
+        # Use a stack per thread per GPU component just in case.
+        self._stack = _GPUStacksLocal()
+
+    def _make_stream(self):
         with self.gpu.cupy_device:
-            self.stream = cupy.cuda.Stream(null=False, non_blocking=True)
+            return cupy.cuda.Stream(null=False, non_blocking=True)
 
     def __enter__(self):
-        self.gpu.cupy_device.__enter__()
-        self.stream.__enter__()
+        _gpu_locals._gpus = self.gpus
+        dev = self.gpu.cupy_device
+        dev.__enter__()
+        stream = self._make_stream()
+        stream.__enter__()
+        self._stack.push_stream(stream)
+        self._stack.push_device(dev)
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
-        self.stream.__exit__(exc_type, exc_val, exc_tb)
-        self.stream.synchronize()
-        return self.gpu.cupy_device.__exit__(exc_type, exc_val, exc_tb)
+        stream = self._stack.pop_stream()
```

Comment:
I think there's an ordering issue here. Shouldn't we synchronize the stream, then exit it, then pop the stream, then pop the device.